Proposal: Choose network at build-time via flags

Anviking opened the pull request:
# Issue Number

#95 

# Motivation
We currently have a `--network` option in the CLI. We could derive the protocol magic (which we need), from this, and pass it around "everywhere". @KtorZ suggested we instead either:

- Choose network at via a compile-time flag (like this)
- Use `unsafePerformIO` reading a ENV var (we could either default to testnet or to nothing)

# Overview

- [x] I have added `config/staging`, `config/testnet`, `config/mainnet` directories
- [x] I have added `Cardano.Network` inside each directory, exporting `protocolMagic :: Int32`
- [x] I have added `mainnet` and `staging` flags and made the `testnet` config default

```
$ stack ghci cardano-wallet
λ> import Cardano.Network
λ> Cardano.Network.protocolMagic
1097911063
```
```
$ stack ghci cardano-wallet --flag cardano-wallet:mainnet
λ> import Cardano.Network
λ> Cardano.Network.protocolMagic
764824073
```


# Comments

- I wonder how this plays out with our integration tests. I do not know if we would like to specify `local` as network, but if we want to, it might be tricky. It appears cabal doesn't support specifying flags, but [stack does](https://stackoverflow.com/a/35123917)
- This commit does not replace existing CLI logic for choosing network,
which would be a natural follow-up, if we were to go down this path.

<!-- 



Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

akegalj commented on src/Cardano/Wallet/Kernel/Ed25519Bip44.hs:
```diff
@@ -1,26 +1,61 @@
 import           Pos.Crypto (PublicKey (..))
 
-import           Cardano.Crypto.Wallet (DerivationScheme (..), deriveXPub)
+import           Cardano.Crypto.Wallet (DerivationScheme (DerivationScheme2),
```
added explicit `DerivationScheme2` so we not confuse things

Anviking committed:
'mainnet' and 'staging' build flags & default to testnet

This commit does not replace existing CLI logic for choosing network,
which would be a natural follow-up, if we were to go down this path.

Anviking commented on cardano-wallet.cabal:
```diff
@@ -81,9 +90,16 @@ library
+      hs-source-dirs: config/mainnet
+  if flag(staging)
+      hs-source-dirs: config/staging
+  if (!flag(staging) && !flag(mainnet))
```
Could go for CPP within one single module, instead of having 3 different files, I guess

KtorZ replied:
Hmmm... Yes, it'll be better to have CPP pragmas within one single file, and changing the local declaration using this. Also, I'd definitely have the testnet flag too, (that is enabled by default).

Anviking commented on cardano-wallet.cabal:
```diff
@@ -31,6 +39,13 @@ library
+  if flag(staging)
+    cpp-options: -DCARDANO_NETWORK_STAGING
+  if (!flag(staging) && !flag(mainnet))
+    cpp-options: -DCARDANO_NETWORK_TESTNET
```
> Also, I'd definitely have the testnet flag too, (that is enabled by default).

@KtorZ That seems tricky as we'd have to turn it off when passing `flag(mainnet)`.

(Couldn't comment on the old thread. I have used CPP now though)

Anviking commented on src/Cardano/Network.hs:
```diff
@@ -0,0 +1,41 @@
+networkName = "testnet"
+
+#endif
+#ifdef CARDANO_NETWORK_STAGING
```
This `#ifdef`/`#endif` was the best I got working.

When passing two flags we get `Multiple declarations of ‘protocolMagic’` errors… which seems good. Picking one over the other could be dangerous.
